### ISSUE TITLE
Extend FlutterActivity with new v2 embedding

### ISSUE THREAD

**Issue by vitor-gyant** [flutter/flutter#43909]
According to the wiki information, the new embedding allows FlutterActivity to be extended in order to provide a cached engine.

```
public class ChatActivity extends FlutterActivity {

    // Override to provide our warmed up engine instance.
    @Nullable
    @Override
    public FlutterEngine provideFlutterEngine(@NonNull Context context) {
        return MyFlutterEngine.getInstance(context);
    }
}

```

However, launching this activity with:

```
        Intent intent = new Intent(this, ChatActivity.class);
        startActivity(intent);
```

Results in a white screen. It works if I launch FlutterActivity directly using:

```
Intent defaultFlutter = new FlutterActivity.createDefaultIntent(currentActivity);
startActivity(defaultFlutter);
```

Is there any additional step to be done ? I'm using the flutter master channel.

Thanks,

**Comment by iapicca** [comment]
Hi @vitor-gyant 
can you please provide your `flutter doctor -v`?
Also, to better address the issue, would be helpful
if you could post a self contained app on github
or the steps to reproduce it.
Thank you

https://github.com/flutter/flutter/issues/43220
https://github.com/flutter/flutter/issues/43921

**Comment by vitor-gyant** [comment]
@iapicca Here is the flutter doctor output.

```
[✓] Flutter (Channel master, v1.10.15-pre.384, on Mac OS X 10.14.6 18G103, locale en-GB)
    • Flutter version 1.10.15-pre.384 at /Users/vmagalhaes/flutter
    • Framework revision ee032f67c7 (3 days ago), 2019-11-02 09:46:25 -0700
    • Engine revision 8ea19b1c76
    • Dart version 2.6.0 (build 2.6.0-dev.8.2 bbe2ac28c9)
```

I could send a sample project I just need to arrange some time. Nevertheless, if you simply create a new project with:

`flutter create -t module --androidx `

You could easily replicate the issue with the ChatActivity class above.

Thanks,

**Comment by iapicca** [comment]
Hi again @vitor-gyant,
I just noticed this code in your first post 

```
public class ChatActivity extends FlutterActivity {

    // Override to provide our warmed up engine instance.
    @Nullable
    @Override
    public FlutterEngine provideFlutterEngine(@NonNull Context context) {
        return MyFlutterEngine.getInstance(context);
    }
}
```

I don't see your code in all it's extent, 
but I'm not sure it's implemented following the [documentation](https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects),
can you please double check?

https://github.com/flutter/flutter/issues/44171
https://github.com/flutter/flutter/issues/44161

**Comment by vitor-gyant** [comment]
@iapicca I've double checked it. My `ChatActivity` class extends `FlutterActivity` and implements `provideFlutterEngine` to return a pre-warmed engine just like the docs state. 

The engine is created and setup before the activity is displayed.

**Comment by iapicca** [comment]
@vitor-gyant  thank you for checking this

**Comment by ekasetiawans** [comment]
> @iapicca I've double checked it. My `ChatActivity` class extends `FlutterActivity` and implements `provideFlutterEngine` to return a pre-warmed engine just like the docs state.
> 
> The engine is created and setup before the activity is displayed.

Have you tried like this :
1. Pre-warm your flutterEngine
2. Put your flutterEngine to `FlutterEngineCache` ex: `FlutterEngineCache.getInstance().put("my_engine", flutterEngine)`
3. Override `getCacheEngineId()` in your Activity that extends FlutterActivity, and return with `"my_engine"` value. ex : 
```
 @Override
   public String getCachedEngineId() {
      return "my_engine";
    }
```

**Comment by scostagabriel** [comment]
I am having the same issue. The only difference is that I'm getting a black screen and in my code is that I'm using one of these two methods to get my cached engine that I have set in MyApplication but both have the same result.

``
FlutterEngine flutterEngine = ((MyApplication) getApplication()).getFlutterEngine();
FlutterEngine flutterEngine  = FlutterEngineCache.getInstance().get("cachedEngine");
``

When I use the method provideFlutterEngine() the logs that I get are below:

`D/FlutterActivityAndFragmentDelegate(11059): Detaching FlutterEngine from the Activity that owns this Fragment.
D/FlutterEngine(11059): Destroying.
D/FlutterEnginePluginRegistry(11059): Destroying.
D/FlutterActivity(11059): Using the launch theme as normal theme.
D/FlutterActivityAndFragmentDelegate(11059): Setting up FlutterEngine.
I/teste   (11059): provideFlutterEngine
D/FlutterActivityAndFragmentDelegate(11059): Attaching FlutterEngine to the Activity that owns this Fragment.
D/FlutterView(11059): Attaching to a FlutterEngine: io.flutter.embedding.engine.FlutterEngine@5ac763
D/FlutterView(11059): Detaching from a FlutterEngine: io.flutter.embedding.engine.FlutterEngine@5ac763`

And when I don't use it (just use the getDartEntrypointFunctionName method in my Activity). The logs are:

`D/FlutterActivityAndFragmentDelegate( 4591): Detaching FlutterEngine from the Activity that owns this Fragment.
D/FlutterEngine( 4591): Destroying.
D/FlutterEnginePluginRegistry( 4591): Destroying.
D/FlutterActivity( 4591): Using the launch theme as normal theme.
D/FlutterActivityAndFragmentDelegate( 4591): Setting up FlutterEngine.
D/FlutterActivityAndFragmentDelegate( 4591): No preferred FlutterEngine was provided. Creating a new FlutterEngine for this FlutterFragment.
D/eglCodecCommon( 4591): setVertexArrayObject: set vao to 0 (0) 0 0
D/EGL_emulation( 4591): eglCreateContext: 0xdee27380: maj 3 min 1 rcv 4
D/HostConnection( 4591): HostConnection::get() New Host Connection established 0xd372d210, tid 4689
D/HostConnection( 4591): HostComposition ext ANDROID_EMU_CHECKSUM_HELPER_v1 ANDROID_EMU_native_sync_v2 ANDROID_EMU_native_sync_v3 ANDROID_EMU_dma_v1 ANDROID_EMU_direct_mem ANDROID_EMU_host_composition_v1 ANDROID_EMU_host_composition_v2 ANDROID_EMU_vulkan ANDROID_EMU_deferred_vulkan_commands ANDROID_EMU_vulkan_null_optional_strings ANDROID_EMU_vulkan_create_resources_with_requirements ANDROID_EMU_YUV420_888_to_NV21 ANDROID_EMU_YUV_Cache GL_OES_EGL_image_external_essl3 GL_OES_vertex_array_object GL_KHR_texture_compression_astc_ldr ANDROID_EMU_gles_max_version_3_1
D/EGL_emulation( 4591): eglMakeCurrent: 0xdee27380: ver 3 1 (tinfo 0xe9f892f0)
E/eglCodecCommon( 4591): glUtilsParamSize: unknow param 0x000082da
E/eglCodecCommon( 4591): glUtilsParamSize: unknow param 0x000082da
D/FlutterActivityAndFragmentDelegate( 4591): Attaching FlutterEngine to the Activity that owns this Fragment.
I/teste   ( 4591): configureFlutterEngineLockScreenActivity2
I/teste   ( 4591): onCreateLockScreenActivity2
D/FlutterView( 4591): Attaching to a FlutterEngine: io.flutter.embedding.engine.FlutterEngine@2798551
I/teste   ( 4591): getDartEntrypointFunctionName
D/FlutterActivityAndFragmentDelegate( 4591): Executing Dart entrypoint: mainStandalone, and sending initial route: /
I/teste   ( 4591): getDartEntrypointFunctionName
D/FlutterView( 4591): Detaching from a FlutterEngine: io.flutter.embedding.engine.FlutterEngine@2798551
I/Choreographer( 4591): Skipped 42 frames!  The application may be doing too much work on its main thread.`

I have also noticed that if I put the method getCachedEngineId() in the Activity the code inside provideFlutterEngine doesn't run.

Here's my flutter doctor -v

`[√] Flutter (Channel stable, v1.12.13+hotfix.8, on Microsoft Windows [versÃ£o 10.0.18362.657], locale pt-BR)
    • Flutter version 1.12.13+hotfix.8 at C:\flutter
    • Framework revision 0b8abb4724 (3 weeks ago), 2020-02-11 11:44:36 -0800
    • Engine revision e1e6ced81d
    • Dart version 2.7.0

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at C:\Users\gabri\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)
    • All Android licenses accepted.

[√] Android Studio (version 3.5)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 41.1.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b03)

[√] VS Code (version 1.42.1)
    • VS Code at C:\Users\gabri\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.8.1

[√] Connected device (1 available)
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 10 (API 29) (emulator)`
